### ISSUE TITLE
The Figure1 problem.

### ISSUE THREAD

**Issue by zhangxiaodu** [Paralysis/ringloss#1]
Hello, I have achieved ringloss as similar as yours. I already found the ringloss is not stable and I finetune my model using ringloss. However, when I draw feature with 2 dim, I can't get the result as Figure1. Is it ok with your achievement?

**Comment by Paralysis** [comment]
Are you referring to MNIST experiments? Then yes, I am able to generate similar features like the paper showed. Can you include your figure so I can get some idea of what happened?

**Comment by zhangxiaodu** [comment]
@Paralysis I have achieved it again and I got the same features as the paper. But the accuracy decreased compared with softmax. I use learning rate 0.01 for softmax and 0.001 for ringloss finetune. And the loss weight is 0.1 for ringloss (case the 1.0 not converge). 
Since I don't know how to add a figure to the comment, could you please contact with me using e-mail? My address is 541010829@qq.com. Thank you.

**Comment by zhangxiaodu** [comment]
@Paralysis Yes, I used 2d features to calculate accuracy. I will try higher features. But the 2d visualization I achieved is something different with the author. For ringloss, all the classes in the paper have same angle range. But for me, different classes have different angle range, e.g class 2 has larger range of angle than class 1? Is it something wrong? The attachments are the figures for my achievement.

**Comment by Paralysis** [comment]
@zhangxiaodu I can't seem to find your figure. Here are some example plots for the MNIST features: [training_mnist.pdf](https://github.com/Paralysis/ringloss/files/2018801/training_mnist.pdf)

**Comment by zhangxiaodu** [comment]
I have a stupid question. The figure you draw uses 5d or other high features to train and then draw 2d figures using like PCA to reduce dimension? Or like me using 2d both on training and drawing 2d figures?

So sorry I have not understood the method of drawing this figure. 

发自我的 iPhone

**Comment by Paralysis** [comment]
@zhangxiaodu No, training of 2D and 5D (even more) are independent. They are two different models. 2D is only used for visualization and others are for performance check.

**Comment by zhangxiaodu** [comment]
@Paralysis Yes, I use 2d for visualization. But I can't get the result like you. I use Sphereface64 model, and initial learning rate is 0.001 finetuned from softmax-based Sphereface64 model. The code I use is as follows. Are the parameters I used wrong?

nnloss = nn.CrossEntropyLoss()
ringloss = RingLoss(type='auto', loss_weight=1.0)
criterion = [nnloss, ringloss]

params = lr_init(model, LR_INITIAL)
optimizer4nn = optim.SGD(params, momentum=0.9)
scheduler = lr_scheduler.StepLR(optimizer4nn, step_size=20, gamma=0.1)

optimizer4ring = optim.SGD(ringloss.parameters(), lr=1.0)
scheduler4ring = lr_scheduler.StepLR(optimizer4ring, step_size=10, gamma=0.1)

optimizer = [optimizer4nn, optimizer4ring]
scheduler = [scheduler, scheduler4ring]

train_model(model, criterion, optimizer, scheduler, data_loaders, dataset_sizes, num_epochs=80)

**Comment by Paralysis** [comment]
@zhangxiaodu For my implementation, I'm using LeNet for MNIST experiments just to make life easier. The paper never talked about what architecture they used for MNIST experiments, but my LeNet result seems very close to their result. Personally I don't think they used a complicated architecture like Sphereface64 for MNIST training. Try LeNet and train with its default settings first.
